### ISSUE TITLE
Decorator holds (Redis) connection for too long

### ISSUE THREAD

**Issue by dmzkrsk** [argaen/aiocache#260]
Current (0.6.0) decorator implementation

```python
    async def decorator(self, f, *args, **kwargs):
        self._conn = self.cache.get_connection()

        async with self._conn:  # acquire
            key = self.get_cache_key(f, args, kwargs)

            value = await self.get_from_cache(key)  # check the key
            if value is not None:
                return value

            result = await f(*args, **kwargs)  # run the wrapped function

            await self.set_in_cache(key, result)  # set the key

        # release the connection
```
acquires a connection, checks the key, runs the wrapped function, sets the key and only then releases the connection

In my case `f` is `aiohttp` request, that takes some time (about 4 seconds) and I run them in batches (20 requests per batch)

All this time the connection is hold, but unused and the connection pool quickly empties, so other connections (which only try to get a value from the cache) hang.

And I get timeout errors:

```
Traceback (most recent call last):
  File ".../lib/python3.6/site-packages/aiocache/decorators.py", line 108, in get_from_cache
    value = await self.conn.get(key)
  File ".../lib/python3.6/site-packages/aiocache/base.py", line 499, in _do_inject_conn
    return await getattr(self._cache, cmd_name)(*args, _conn=self._conn, **kwargs)
  File ".../lib/python3.6/site-packages/aiocache/base.py", line 56, in _enabled
    return await func(*args, **kwargs)
  File ".../python3.6/site-packages/aiocache/base.py", line 41, in _timeout
    return await asyncio.wait_for(func(self, *args, **kwargs), timeout)
  File "/usr/local/Cellar/python3/3.6.0/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/tasks.py", line 356, in wait_for
    raise futures.TimeoutError()
```

Increasing `pool_max_size` helps, but I think it's wrong.

I tried to refactor `decorator` method, so connection is hold only for get/set operations:

```python
    async def decorator(self, f, *args, **kwargs):
        key = self.get_cache_key(f, args, kwargs)

        self._conn = self.cache.get_connection()
        async with self._conn:
            value = await self.get_from_cache(key)

        if value is not None:
            return value

        result = await f(*args, **kwargs)

        self._conn = self.cache.get_connection()  # should get_connection twice
        async with self._conn:
            await self.set_in_cache(key, result)

        return result
```

And it works much better.

**Comment by argaen** [comment]
True, this behavior is not correct. Instead of what you are proposing, we could use directly `self.cache.set` and `self.cache.get` for the needed operations. Right now there is no advantage on acquiring the connection for executing, we can leave the cache to do it.

Are you OK providing a MR with the changes? I can do a small release after it :)

**Comment by dmzkrsk** [comment]
Also in `self._conn = self.cache.get_connection()` `self._conn` is rewritten by all my requests and `get_cache_key` uses the same (wrong) connection eventually.

But using `self.cache.set` will fix it as well.

Also, do I need to move

```python
            if value is not None:
                asyncio.ensure_future(self.cache.close())
```
from `get_from_cache` to `decorator`?

I will make some test on my project and submit a MR a bit later

**Comment by argaen** [comment]
The close shouldn't be there, I think I forgot to remove it with the latest refactor I did. I will do the change for that thing and add a couple more tests this afternoon

You can do the change to use the `cache` instead of `_conn` and provide the MR for that if its OK.

**Comment by argaen** [comment]
Closed by #267
